src/Features/Core/Completion/AbstractCompletionService.cs

```diff
@@ -8,94 +8,57 @@
 
 namespace Microsoft.CodeAnalysis.Completion
 {
-    internal abstract partial class AbstractCompletionService : ICompletionService, ITextCompletionService
+    internal abstract partial class AbstractCompletionService : ICompletionService
     {
         private static readonly Func<string, List<CompletionItem>> s_createList = _ => new List<CompletionItem>();
 
-        private const int MruSize = 10;
-
-        private readonly List<string> _committedItems = new List<string>(MruSize);
-        private readonly object _mruGate = new object();
-
-        internal void CompletionItemCommitted(CompletionItem item)
-        {
-            lock (_mruGate)
-            {
-                // We need to remove the item if it's already in the list.
-                // If we're at capacity, we need to remove the LRU item.
-                var removed = _committedItems.Remove(item.DisplayText);
-                if (!removed && _committedItems.Count == MruSize)
-                {
-                    _committedItems.RemoveAt(0);
-                }
-
-                _committedItems.Add(item.DisplayText);
-            }
-        }
+        public string LanguageName { get; }
+        public MostRecentlyUsedList MostRecentlyUsedList { get; } = new MostRecentlyUsedList();
 
         internal int GetMRUIndex(CompletionItem item)
         {
-            lock (_mruGate)
-            {
-                // A lower value indicates more recently used.  Since items are added
-                // to the end of the list, our result just maps to the negation of the 
-                // index.
-                // -1 => 1  == Not Found
-                // 0  => 0  == least recently used 
-                // 9  => -9 == most recently used 
-                var index = _committedItems.IndexOf(item.DisplayText);
-                return -index;
-            }
+            return this.MostRecentlyUsedList.GetMRUIndex(item);
         }
 
         public void ClearMRUCache()
         {
-            lock (_mruGate)
-            {
-                _committedItems.Clear();
-            }
+            this.MostRecentlyUsedList.Clear();
         }
 
-        /// <summary>
-        /// Apply any culture-specific quirks to the given text for the purposes of pattern matching.
-        /// For example, in the Turkish locale, capital 'i's should be treated specially in Visual Basic.
-        /// </summary>
-        public virtual string GetCultureSpecificQuirks(string candidate)
+        protected AbstractCompletionService(string languageName)
         {
-            return candidate;
+            this.LanguageName = languageName;
         }
 
         public abstract IEnumerable<CompletionListProvider> GetDefaultCompletionProviders();
 
-        protected abstract string GetLanguageName();
+        protected virtual CompletionRules CreateCompletionRules(MostRecentlyUsedList mostRecentlyUsedList)
+        {
+            return new CompletionRules(mostRecentlyUsedList);
```

Comment:
Yup -- I've got a bug to unify the commit characters. I also noticed that we create a new ```CompletionRules``` every time GetCompletionRules() is called, which doesn't seem right since there's very little state in the rules. The only thing that isn't stateless in the rules is a cache of pattern matchers, which I may move elsewhere. As an aside, it's looking like we'll need to make ```PatternMatcher``` public with the completion API. Otherwise, implementing CompletionRules is pretty tough if you want custom behavior.